src/github.com/matrix-org/dendrite/encryptoapi/routing/keys.go

```diff
@@ -0,0 +1,537 @@
+// whether device or one time or both of them
+func checkUpload(req *types.UploadEncryptSpecific, typ int) bool {
+	if typ == BODYDEVICEKEY {
+		devicekey := req.DeviceKeys
+		if devicekey.UserID == "" {
+			return false
+		}
+	}
+	if typ == BODYONETIMEKEY {
+		if req.OneTimeKey.KeyString == nil || req.OneTimeKey.KeyObject == nil {
+			return false
+		}
+	}
+	return true
+}
+
+// QueryOneTimeKeys todo: complete this field through claim type
+func QueryOneTimeKeys(
+	ctx context.Context,
+	typ int,
+	userID, deviceID string,
+	encryptionDB *storage.Database,
+) interface{} {
+	if typ == TYPESUM {
+		res, _ := encryptionDB.SelectOneTimeKeyCount(ctx, deviceID, userID)
+		return res
+	}
+	return nil
+}
+
+// ClearUnused when web client sign out, a clean should be processed, cause all keys would never been used from then on.
+// todo: complete this function and invoke through sign out extension or some scenarios else those matter
+func ClearUnused() {}
+
+// persist both device keys and one time keys
+func persistKeys(
+	ctx context.Context,
+	database *storage.Database,
+	body *types.UploadEncryptSpecific,
+	userID,
+	deviceID string,
+) (err error) {
+	// in order to persist keys , a check filtering duplicate should be processed
+	// true stands for counterparts are in request
+	// situation 1: only device keys
+	// situation 2: both device keys and one time keys
+	// situation 3: only one time keys
+	if checkUpload(body, BODYDEVICEKEY) {
+		deviceKeys := body.DeviceKeys
+		al := deviceKeys.Algorithm
+		err = persistAl(ctx, *database, userID, deviceID, al)
+		if err != nil {
+			return
+		}
+		if checkUpload(body, BODYONETIMEKEY) {
+			if err = bothKeyProcess(ctx, body, userID, deviceID, database, deviceKeys); err != nil {
+				return
+			}
+		} else {
+			if err = dkeyProcess(ctx, userID, deviceID, database, deviceKeys); err != nil {
+				return
+			}
+		}
+		// notifier to sync server
+		upnotify(userID)
+	} else {
+		if checkUpload(body, BODYONETIMEKEY) {
+			if err = otmKeyProcess(ctx, body, userID, deviceID, database); err != nil {
+				return
+			}
+		} else {
+			return errors.New("failed to touch keys")
+		}
+	}
+	return err
+}
+
+// make keys instantiated to specific struct from keybody interface{}
+func turnSpecific(
+	cont types.UploadEncrypt,
+) (spec types.UploadEncryptSpecific) {
+	// both device keys are coordinate
+	spec.DeviceKeys = cont.DeviceKeys
+	spec.OneTimeKey.KeyString = make(map[string]string)
+	spec.OneTimeKey.KeyObject = make(map[string]types.KeyObject)
+	mapStringInterface := cont.OneTimeKey
+	for key, val := range mapStringInterface {
+		value, ok := val.(string)
+		if ok {
+			spec.OneTimeKey.KeyString[key] = value
+		} else {
+			valueObject := types.KeyObject{}
+			target, _ := json.Marshal(val)
+			err := json.Unmarshal(target, &valueObject)
+			if err != nil {
+				continue
+			}
+			spec.OneTimeKey.KeyObject[key] = valueObject
+		}
+	}
+	return
+}
+
+func persistAl(
```

Comment:
It's no that complicated , but what if implementation changed someday as e2e method changing , I can still invoke persistAl , not to care about it's implementation , it may look a verbose one today but not all the same in the future.